### DISCUSSION TITLE
Request cancelled in http client send()

### DISCUSSION THREAD

**Discussion by mrAndersen** [reactphp/reactphp#590] [Q&A]
I am getting RuntimeException with message "Request cancelled", and no underlying information about why it was cancelled. How I can get that information?

The code is simply:
```php
try {
            /** @var ResponseInterface $response */
            $response = await(
                $browser
                    ->withResponseBuffer(1024 * 1024 * 32)
                    ->withTimeout($timeout)
                    ->request($method, $url, $headers, $body)
            );
} catch (Throwable $throwable) {
            $this->httpClientLogger->critical($throwable->getMessage(), [
                'url' => $url,
                'elapsed' => round(($end - $start) * 1000, 2),
            ]);

            return null;
        }
```

**Comment by WyriHaximus** [comment]
You get that error when you cancel the promise returned by `Browser::request`.

**Reply by mrAndersen** [reply]
Nope, I am not cancelling manually anything. However I figured out what is going on.

If underlying HTTP request which is made to backend that is running via react-php is being cancelled (for example because of timeout), IF in that case that react-php backend is processing any other http request by itself to some third party remote, that http request which was started by react-php backed will be cancelled with the following message, because the underlying http request to that react-php backend is being cancelled. 

Reproduce for this one is fairly easy, but strictly speaking it's not a bug, just the message should be more specific, so I guess it should be separated to "Request canceled manually" and "Request cancelled because of external promise rejection" or something similar.

**Reply by clue** [reply]
@mrAndersen I'm not sure I follow what problem you're seeing exactly, can you provide a simple gist to reproduce this? Ideally something I could run without external requirements, as the above snippet contains a number of unknown variables and components.

**Reply by mrAndersen** [reply]
@clue That's not a problem at all, I am just saying that when some particular http request that was sent to backend which is managed by react-php is being cancelled, if at the same time this react-php backend is itself sending http request on the same loop to some other remote, then this http request (or several http requests) will be cancelled with that message "Request cancelled", even if inside the code there was no signs of calling reject manually which is a bit confusing. And because of that I spent a bit of time to pin point the reason why requests are being cancelled. The actual logic is fine

So I proposed to change naming a bit and separate the situation when promise was cancelled manually, or when it was cancelled because of other reasons. 

I will make reproduce in couple of hours

**Reply by mrAndersen** [reply]
@clue Here is reproduce https://github.com/mrAndersen/react-php-repr-cancel

This will start HTTP server with one route that will try to do http request during it's execution. So if you run node req.js (which will close connection after only 100ms), you will see in server logs Uncaught PHP Exception RuntimeException: "Request cancelled"

Which is not very clear as I said before, on big projects it can be quite difficult to figure out what request and where was cancelled, and why. So as I said before it would be nice to have some separate messages for that.


P.S.

Imagine some kind of backend that is doing a lot of different http requests, some are 500s, some 400s, some unable to resolve DNS, and e.t.c, and when you are getting just "Request cancelled" during that flow of errors, you are a bit confused, why cancelled, when, by whom, e.t.c.
